File: php/13-罗马数字转整数.php
<?php

class Solution {

    /**
     * @param String $s
     * @return Integer
     */
    function romanToInt($s) {
        $arr = ['I' => 1, 'V' => 5, 'X' => 10, 'L' => 50, 'C' => 100, 'D' => 500, 'M' => 1000];
        $a = str_split($s);

        $sum = 0;
        $temp = 0; //记录上一个值
        
        foreach ($a as $value) {
            if ($temp && $arr[$value] > $temp) {
                $sum += $arr[$value] - $temp * 2; //把上次多加的减掉
            } else {
                $sum += $arr[$value];
            }
            $temp = $arr[$value];
        }
        return $sum;
    }
}

File: c++/13-罗马数字转整数.cpp
class Solution {
public:
    int romanToInt(string s) {
        unordered_map<string, int>mmap = {{"I", 1}, {"V", 5}, {"X", 10}, {"L", 50}, 
        {"C", 100}, {"D", 500}, {"M", 1000}};

        int temp = 0;
        int sum = 0;
        for (int i = 0; i < s.size(); i++) {
            int num = mmap[s.substr(i, 1)];
            if( temp &&  num > temp ) {
                sum += num - 2 * temp;
            } else {
                sum += num;
            }
            temp = num;
        }

        return sum;
    }
};